File: mobile/lib/core/widgets/editor/link_utils.dart
import 'package:flutter/material.dart';
import 'package:flutter_quill/flutter_quill.dart';
import 'package:url_launcher/url_launcher.dart';

import '../app_snackbar.dart';

const _schemes = [
  'http://',
  'https://',
  'mailto:',
  'tel:',
  'sms:',
  'ftp://',
  'ftps://',
  'file://',
  'geo:',
];

final _hostLike = RegExp(
  r'^([\w-]+(\.[\w-]+)+|localhost)(:\d+)?([/?#].*)?$',
  caseSensitive: false,
);

String normalizeUrl(String raw) {
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return trimmed;
  final lower = trimmed.toLowerCase();
  for (final scheme in _schemes) {
    if (lower.startsWith(scheme)) return trimmed;
  }
  if (_hostLike.hasMatch(trimmed)) return 'https://$trimmed';
  return trimmed;
}

bool isLikelyUrl(String raw) {
  final trimmed = raw.trim();
  if (trimmed.isEmpty || trimmed.contains(' ')) return false;
  final lower = trimmed.toLowerCase();
  for (final scheme in _schemes) {
    if (lower.startsWith(scheme)) return true;
  }
  return _hostLike.hasMatch(trimmed);
}

Future<void> launchExternal(BuildContext context, String url) async {
  final normalized = normalizeUrl(url);
  final uri = Uri.tryParse(normalized);
  if (uri == null) {
    if (context.mounted) {
      AppSnackbar.showError(context, message: 'Invalid link');
    }
    return;
  }
  try {
    final ok = await launchUrl(uri, mode: LaunchMode.externalApplication);
    if (!ok && context.mounted) {
      AppSnackbar.showError(context, message: "Couldn't open link");
    }
  } catch (_) {
    if (context.mounted) {
      AppSnackbar.showError(context, message: "Couldn't open link");
    }
  }
}

({String url, String text, int start, int length})? linkAtSelection(
  QuillController controller,
) {
  final sel = controller.selection;
  if (!sel.isValid || sel.baseOffset < 0) return null;
  final offset = sel.baseOffset;

  final ops = controller.document.toDelta().toList();
  var pos = 0;
  var hitIndex = -1;
  String? hitUrl;
  String hitText = '';

  for (var i = 0; i < ops.length; i++) {
    final op = ops[i];
    final data = op.data;
    if (data is! String) {
      pos += op.length ?? 0;
      continue;
    }
    if (offset >= pos && offset <= pos + data.length) {
      final url = op.attributes?['link'] as String?;
      if (url != null && url.isNotEmpty) {
        hitIndex = i;
        hitUrl = url;
        hitText = data;
        break;
      }
      if (offset < pos + data.length) return null;
    }
    pos += data.length;
  }

  if (hitIndex < 0 || hitUrl == null) return null;

  var start = pos;
  final left = <String>[];
  for (var i = hitIndex - 1; i >= 0; i--) {
    final data = ops[i].data;
    if (data is! String) break;
    if ((ops[i].attributes?['link'] as String?) != hitUrl) break;
    left.add(data);
    start -= data.length;
  }

  final buffer = StringBuffer()
    ..writeAll(left.reversed)
    ..write(hitText);
  for (var i = hitIndex + 1; i < ops.length; i++) {
    final data = ops[i].data;
    if (data is! String) break;
    if ((ops[i].attributes?['link'] as String?) != hitUrl) break;
    buffer.write(data);
  }

  final text = buffer.toString();
  return (url: hitUrl, text: text, start: start, length: text.length);
}

File: web/features/notes/link-utils.ts
import type { QuillInstance } from "./quill";

const SCHEMES = [
  "http://",
  "https://",
  "mailto:",
  "tel:",
  "sms:",
  "ftp://",
  "ftps://",
  "file://",
  "geo:",
];

const HOST_LIKE = /^(?:[\w-]+(?:\.[\w-]+)+|localhost)(?::\d+)?(?:[/?#].*)?$/i;
const IPV4_HOST = /^(?:\d{1,3}\.){3}\d{1,3}$/;

// Dotted-but-all-numeric strings like "3.14" or "192.168" are prose, not
// hosts; only a full dotted quad counts as a numeric host.
function isHostLike(trimmed: string): boolean {
  if (!HOST_LIKE.test(trimmed)) return false;
  const host = trimmed.split(/[/?#:]/, 1)[0];
  return /[a-z]/i.test(host) || IPV4_HOST.test(host);
}

export function normalizeUrl(raw: string): string {
  const trimmed = raw.trim();
  if (!trimmed) return trimmed;
  const lower = trimmed.toLowerCase();
  for (const scheme of SCHEMES) {
    if (lower.startsWith(scheme)) return trimmed;
  }
  if (isHostLike(trimmed)) return `https://${trimmed}`;
  return trimmed;
}

export function isLikelyUrl(raw: string): boolean {
  const trimmed = raw.trim();
  if (!trimmed || /\s/.test(trimmed)) return false;
  const lower = trimmed.toLowerCase();
  for (const scheme of SCHEMES) {
    if (lower.startsWith(scheme)) return true;
  }
  return isHostLike(trimmed);
}

export type LinkRange = {
  url: string;
  text: string;
  start: number;
  length: number;
};

export function linkAtIndex(
  quill: QuillInstance,
  index: number,
): LinkRange | null {
  if (index < 0) return null;
  const ops = quill.getContents().ops ?? [];
  let pos = 0;
  let hitIndex = -1;
  let hitUrl: string | null = null;
  let hitText = "";

  for (let i = 0; i < ops.length; i++) {
    const data = ops[i].insert;
    if (typeof data !== "string") {
      pos += 1;
      continue;
    }
    if (index >= pos && index <= pos + data.length) {
      const url =
        typeof ops[i].attributes?.link === "string"
          ? (ops[i].attributes?.link as string)
          : null;
      if (url) {
        hitIndex = i;
        hitUrl = url;
        hitText = data;
        break;
      }
      if (index < pos + data.length) return null;
    }
    pos += data.length;
  }

  if (hitIndex < 0 || !hitUrl) return null;

  let start = pos;
  let text = hitText;
  for (let i = hitIndex - 1; i >= 0; i--) {
    const data = ops[i].insert;
    if (typeof data !== "string") break;
    if (ops[i].attributes?.link !== hitUrl) break;
    text = data + text;
    start -= data.length;
  }
  for (let i = hitIndex + 1; i < ops.length; i++) {
    const data = ops[i].insert;
    if (typeof data !== "string") break;
    if (ops[i].attributes?.link !== hitUrl) break;
    text += data;
  }

  return { url: hitUrl, text, start, length: text.length };
}
